File: Top 100 Liked/Arrays/0442_find_all_duplicates_in_an_array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        n = len(nums)
        for i in range(n):
            while nums[i] != nums[nums[i] - 1]:
                nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
        ans = []
        for i in range(n):
            if nums[i] != i + 1:
                ans.append(nums[i])
        return ans

# another in-place solution:
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = []
        for i in range(n):
            tmp = abs(nums[i])
            if nums[tmp - 1] > 0:
                nums[tmp - 1] *= -1
            else:
                ans.append(tmp)
        return ans

File: Top 100 Liked/Arrays/0442_find_all_duplicates_in_an_array.java
// hashset solution:
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        Set<Integer> hashset = new HashSet<>();
        int n = nums.length;
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (!hashset.contains(nums[i])) {
                hashset.add(nums[i]);
            } else {
                ans.add(nums[i]);
            }
        }
        return ans;
    }
}

// in-place hashing solution:
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        int n = nums.length;
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            while (nums[i] != nums[nums[i] - 1]) {
                swap(nums, i, nums[i] - 1);
            }
        }
        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1) {
                ans.add(nums[i]);
            }
        }
        return ans;
    }

    private void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}
